File: apps/desktop/src/core/storage/SettingsStore.ts
import type {
  ChartDisplayPreferences,
  IndicatorId,
  IndicatorSetting,
  IndicatorSettingsState,
} from '@0dtetrader/shared-types';
import {
  DEFAULT_CHART_DISPLAY,
  DEFAULT_INDICATOR_SETTINGS_STATE,
  validateIndicatorSettingsState,
} from '../../features/chart/indicatorRegistry';
import type { OptionsAnalyticsSettings } from '../../features/chart/optionsAnalytics/optionsAnalyticsSettings';
import {
  decodeOptionsAnalyticsSettings,
  DEFAULT_OPTIONS_ANALYTICS_SETTINGS,
} from '../../features/chart/optionsAnalytics/optionsAnalyticsSettings';
import type { ChartTradingSettings } from '../../features/chart/chartTradingSettings';
import {
  decodeChartTradingSettings,
  DEFAULT_CHART_TRADING_SETTINGS,
} from '../../features/chart/chartTradingSettings';
import type { TwcHeatmapSettings } from '../../features/chart/twc/twcSettings';
import { DEFAULT_TWC_SETTINGS } from '../../features/chart/twc/twcSettings';
import type { UsrSettings } from '../../features/chart/ultimateSupportResistance/usrSettings';
import {
  decodeUsrSettings,
  DEFAULT_USR_SETTINGS,
} from '../../features/chart/ultimateSupportResistance/usrSettings';

export type TradeLayout = 'fullscreen' | 'split';

/** 'termStructure': strike x expiration, latest snapshot per expiration.
 *  'timeSeries': strike x timestamp, one expiration over its capture history. */
export type GexHeatmapViewMode = 'termStructure' | 'timeSeries';

/** localStorage-backed app settings (SettingsStore.swift analog). */
export class SettingsStore {
  private static keys = {
    layoutMode: 'settings.layoutMode',
    legacyIndicatorSettings: 'settings.indicatorSettings',
    indicatorSettings: 'settings.indicatorSettings.v1',
    chartDisplay: 'settings.chartDisplay.v1',
    twcSettings: 'settings.twcSettings',
    usrSettings: 'settings.ultimateSupportResistance.v1',
    optionsAnalytics: 'settings.optionsAnalytics.v1',
    chartTrading: 'settings.chartTrading.v1',
    riskDisclaimerAccepted: 'settings.riskDisclaimerAccepted',
    serverSelectionCompleted: 'settings.serverSelectionCompleted',
    lastSymbol: 'settings.lastSymbol',
    tradingLocked: 'settings.tradingLocked',
    bypassOrderConfirmation: 'settings.bypassOrderConfirmation',
    keyboardShortcutsEnabled: 'settings.keyboardShortcutsEnabled',
    toastsEnabled: 'settings.toastsEnabled',
    systemNotificationsEnabled: 'settings.systemNotificationsEnabled',
    gexHeatmapView: 'settings.gexHeatmapView',
  };

  get layoutMode(): TradeLayout {
    const stored = localStorage.getItem(SettingsStore.keys.layoutMode);
    return stored === 'fullscreen' || stored === 'split' ? stored : 'split';
  }

  set layoutMode(value: TradeLayout) {
    localStorage.setItem(SettingsStore.keys.layoutMode, value);
  }

  /** Term structure is the default: it matches the reference implementation
   *  this feature was modeled on. */
  get gexHeatmapView(): GexHeatmapViewMode {
    const stored = localStorage.getItem(SettingsStore.keys.gexHeatmapView);
    return stored === 'timeSeries' ? 'timeSeries' : 'termStructure';
  }

  set gexHeatmapView(value: GexHeatmapViewMode) {
    localStorage.setItem(SettingsStore.keys.gexHeatmapView, value);
  }

  get indicatorSettings(): IndicatorSettingsState {
    this.ensureIndicatorMigration();
    const raw = localStorage.getItem(SettingsStore.keys.indicatorSettings);
    if (!raw) return structuredClone(DEFAULT_INDICATOR_SETTINGS_STATE);
    try {
      return validateIndicatorSettingsState(JSON.parse(raw), DEFAULT_INDICATOR_SETTINGS_STATE)
        .value;
    } catch {
      return structuredClone(DEFAULT_INDICATOR_SETTINGS_STATE);
    }
  }

  set indicatorSettings(value: IndicatorSettingsState) {
    const result = validateIndicatorSettingsState(value, DEFAULT_INDICATOR_SETTINGS_STATE);
    if (!result.ok) throw new Error(result.error);
    localStorage.setItem(SettingsStore.keys.indicatorSettings, JSON.stringify(result.value));
  }

  get chartDisplay(): ChartDisplayPreferences {
    this.ensureIndicatorMigration();
    return this.readChartDisplay();
  }

  set chartDisplay(value: ChartDisplayPreferences) {
    if (typeof value.volumeEnabled !== 'boolean') throw new Error('Volume display is invalid.');
    if (typeof value.volumeWeightedCandleWidth !== 'boolean') {
      throw new Error('Volume-weighted candle width display is invalid.');
    }
    localStorage.setItem(SettingsStore.keys.chartDisplay, JSON.stringify(value));
  }

  private readChartDisplay(): ChartDisplayPreferences {
    const raw = localStorage.getItem(SettingsStore.keys.chartDisplay);
    if (!raw) return { ...DEFAULT_CHART_DISPLAY };
    try {
      const value = JSON.parse(raw) as Record<string, unknown>;
      if (typeof value.volumeEnabled !== 'boolean') return { ...DEFAULT_CHART_DISPLAY };
      return {
        volumeEnabled: value.volumeEnabled,
        volumeWeightedCandleWidth:
          typeof value.volumeWeightedCandleWidth === 'boolean'
            ? value.volumeWeightedCandleWidth
            : DEFAULT_CHART_DISPLAY.volumeWeightedCandleWidth,
      };
    } catch {
      return { ...DEFAULT_CHART_DISPLAY };
    }
  }

  private ensureIndicatorMigration(): void {
    const current = localStorage.getItem(SettingsStore.keys.indicatorSettings);
    const legacyRaw = localStorage.getItem(SettingsStore.keys.legacyIndicatorSettings);
    if (current) {
      try {
        const validCurrent = validateIndicatorSettingsState(
          JSON.parse(current),
          DEFAULT_INDICATOR_SETTINGS_STATE,
        ).ok;
        if (validCurrent) {
          let displayValid = false;
          try {
            const displayRaw = localStorage.getItem(SettingsStore.keys.chartDisplay);
            if (displayRaw) {
              const display = JSON.parse(displayRaw) as Record<string, unknown>;
              displayValid = typeof display.volumeEnabled === 'boolean';
            }
          } catch {
            displayValid = false;
          }
          if (!displayValid) {
            try {
              localStorage.setItem(
                SettingsStore.keys.chartDisplay,
                JSON.stringify(DEFAULT_CHART_DISPLAY),
              );
              const repaired = JSON.parse(
                localStorage.getItem(SettingsStore.keys.chartDisplay) ?? 'null',
              ) as Record<string, unknown> | null;
              if (repaired?.volumeEnabled !== DEFAULT_CHART_DISPLAY.volumeEnabled) return;
            } catch {
              // Preserve both the valid keyed record and legacy residue for a later retry.
              return;
            }
          }
          if (legacyRaw) localStorage.removeItem(SettingsStore.keys.legacyIndicatorSettings);
          return;
        }
      } catch {
        // A corrupt current record may still be recoverable from the legacy record.
      }
    }
    if (!legacyRaw) return;
    let legacy: Record<string, unknown>;
    try {
      const parsed: unknown = JSON.parse(legacyRaw);
      if (!parsed || typeof parsed !== 'object' || Array.isArray(parsed)) return;
      legacy = parsed as Record<string, unknown>;
    } catch {
      return;
    }
    const recognized = new Set([
      'smaEnabled',
      'smaPeriod',
      'emaEnabled',
      'emaPeriod',
      'vwapEnabled',
      'rsiEnabled',
      'rsiPeriod',
      'macdEnabled',
      'macdFastPeriod',
      'macdSlowPeriod',
      'macdSignalPeriod',
      'bollingerEnabled',
      'bollingerPeriod',
      'bollingerMultiplier',
      'stochEnabled',
      'stochKPeriod',
      'stochKSmooth',
      'stochDPeriod',
      'atrEnabled',
      'atrPeriod',
      'volumeEnabled',
    ]);
    if (!Object.keys(legacy).some((key) => recognized.has(key))) return;

    const candidate = structuredClone(DEFAULT_INDICATOR_SETTINGS_STATE);
    const bool = (key: string, fallback: boolean): boolean | unknown => legacy[key] ?? fallback;
    const num = (key: string, fallback: number): number | unknown => legacy[key] ?? fallback;
    const set = (
      id: IndicatorId,
      enabledKey: string | null,
      parameterMap: Record<string, string>,
    ) => {
      const defaults = candidate.indicators[id];
      candidate.indicators[id] = {
        enabled: (enabledKey ? bool(enabledKey, defaults.enabled) : defaults.enabled) as boolean,
        parameters: Object.fromEntries(
          Object.entries(defaults.parameters).map(([parameterId, fallback]) => [
            parameterId,
            parameterMap[parameterId] ? num(parameterMap[parameterId], fallback) : fallback,
          ]),
        ) as Record<string, number>,
      } as IndicatorSetting;
    };
    set('sma', 'smaEnabled', { period: 'smaPeriod' });
    set('ema', 'emaEnabled', { period: 'emaPeriod' });
    set('anchored_vwap', 'vwapEnabled', {});
    set('rsi', 'rsiEnabled', { period: 'rsiPeriod' });
    set('macd', 'macdEnabled', {
      fastPeriod: 'macdFastPeriod',
      slowPeriod: 'macdSlowPeriod',
      signalPeriod: 'macdSignalPeriod',
    });
    set('bollinger', 'bollingerEnabled', {
      period: 'bollingerPeriod',
      multiplier: 'bollingerMultiplier',
    });
    set('stochastic', 'stochEnabled', {
      kPeriod: 'stochKPeriod',
      kSmooth: 'stochKSmooth',
      dPeriod: 'stochDPeriod',
    });
    set('atr', 'atrEnabled', { period: 'atrPeriod' });
    const validated = validateIndicatorSettingsState(candidate, DEFAULT_INDICATOR_SETTINGS_STATE);
    const volumeEnabled = bool('volumeEnabled', DEFAULT_CHART_DISPLAY.volumeEnabled);
    if (!validated.ok || typeof volumeEnabled !== 'boolean') return;
    const chartDisplay = { volumeEnabled };

    const previousSettings = localStorage.getItem(SettingsStore.keys.indicatorSettings);
    const previousDisplay = localStorage.getItem(SettingsStore.keys.chartDisplay);
    try {
      localStorage.setItem(SettingsStore.keys.indicatorSettings, JSON.stringify(validated.value));
      localStorage.setItem(SettingsStore.keys.chartDisplay, JSON.stringify(chartDisplay));
      const rereadSettings = localStorage.getItem(SettingsStore.keys.indicatorSettings);
      const rereadDisplay = localStorage.getItem(SettingsStore.keys.chartDisplay);
      if (
        !rereadSettings ||
        !validateIndicatorSettingsState(
          JSON.parse(rereadSettings),
          DEFAULT_INDICATOR_SETTINGS_STATE,
        ).ok ||
        !rereadDisplay ||
        (JSON.parse(rereadDisplay) as { volumeEnabled?: unknown }).volumeEnabled !== volumeEnabled
      ) {
        throw new Error('Indicator settings migration verification failed.');
      }
      localStorage.removeItem(SettingsStore.keys.legacyIndicatorSettings);
    } catch {
      if (previousSettings === null) localStorage.removeItem(SettingsStore.keys.indicatorSettings);
      else localStorage.setItem(SettingsStore.keys.indicatorSettings, previousSettings);
      if (previousDisplay === null) localStorage.removeItem(SettingsStore.keys.chartDisplay);
      else localStorage.setItem(SettingsStore.keys.chartDisplay, previousDisplay);
    }
  }

  get twcSettings(): TwcHeatmapSettings {
    const raw = localStorage.getItem(SettingsStore.keys.twcSettings);
    if (!raw) return DEFAULT_TWC_SETTINGS;
    try {
      return { ...DEFAULT_TWC_SETTINGS, ...(JSON.parse(raw) as Partial<TwcHeatmapSettings>) };
    } catch {
      return DEFAULT_TWC_SETTINGS;
    }
  }

  set twcSettings(value: TwcHeatmapSettings) {
    localStorage.setItem(SettingsStore.keys.twcSettings, JSON.stringify(value));
  }

  get usrSettings(): UsrSettings {
    const raw = localStorage.getItem(SettingsStore.keys.usrSettings);
    if (!raw) return { ...DEFAULT_USR_SETTINGS };
    try {
      return decodeUsrSettings(JSON.parse(raw));
    } catch {
      return { ...DEFAULT_USR_SETTINGS };
    }
  }

  set usrSettings(value: UsrSettings) {
    localStorage.setItem(SettingsStore.keys.usrSettings, JSON.stringify(decodeUsrSettings(value)));
  }

  get optionsAnalytics(): OptionsAnalyticsSettings {
    const raw = localStorage.getItem(SettingsStore.keys.optionsAnalytics);
    if (!raw) return DEFAULT_OPTIONS_ANALYTICS_SETTINGS;
    try {
      return decodeOptionsAnalyticsSettings(JSON.parse(raw));
    } catch {
      return DEFAULT_OPTIONS_ANALYTICS_SETTINGS;
    }
  }

  set optionsAnalytics(value: OptionsAnalyticsSettings) {
    localStorage.setItem(SettingsStore.keys.optionsAnalytics, JSON.stringify(value));
  }

  get chartTrading(): ChartTradingSettings {
    const raw = localStorage.getItem(SettingsStore.keys.chartTrading);
    if (!raw) return DEFAULT_CHART_TRADING_SETTINGS;
    try {
      return decodeChartTradingSettings(JSON.parse(raw));
    } catch {
      return DEFAULT_CHART_TRADING_SETTINGS;
    }
  }

  set chartTrading(value: ChartTradingSettings) {
    localStorage.setItem(SettingsStore.keys.chartTrading, JSON.stringify(value));
  }

  get hasAcceptedRiskDisclaimer(): boolean {
    return localStorage.getItem(SettingsStore.keys.riskDisclaimerAccepted) === 'true';
  }

  set hasAcceptedRiskDisclaimer(value: boolean) {
    localStorage.setItem(SettingsStore.keys.riskDisclaimerAccepted, String(value));
  }

  get hasCompletedServerSelection(): boolean {
    if (localStorage.getItem(SettingsStore.keys.serverSelectionCompleted) === 'true') return true;
    if (!this.hasAcceptedRiskDisclaimer) return false;
    for (let index = 0; index < localStorage.length; index += 1) {
      const key = localStorage.key(index);
      if (!key || key === SettingsStore.keys.riskDisclaimerAccepted) continue;
      return true;
    }
    return false;
  }

  set hasCompletedServerSelection(value: boolean) {
    localStorage.setItem(SettingsStore.keys.serverSelectionCompleted, String(value));
  }

  /** Trading lock: when true, every order-placing control is disabled. Persists
   *  across launches (the lock is remembered, like the layout choice). */
  get tradingLocked(): boolean {
    return localStorage.getItem(SettingsStore.keys.tradingLocked) === 'true';
  }

  set tradingLocked(value: boolean) {
    localStorage.setItem(SettingsStore.keys.tradingLocked, String(value));
  }

  /** Skip the buy/sell confirmation sheet and submit immediately. Per-device. */
  get bypassOrderConfirmation(): boolean {
    return localStorage.getItem(SettingsStore.keys.bypassOrderConfirmation) === 'true';
  }

  set bypassOrderConfirmation(value: boolean) {
    localStorage.setItem(SettingsStore.keys.bypassOrderConfirmation, String(value));
  }

  /** Desktop-grid trading hotkeys (B/S/L — Cmd+K symbol search is always on,
   *  see useTradeShortcuts). Defaults off: arming a real order from a single
   *  keystroke should be opt-in, not a surprise the first time a trader
   *  types "b" somewhere unexpected. */
  get keyboardShortcutsEnabled(): boolean {
    const stored = localStorage.getItem(SettingsStore.keys.keyboardShortcutsEnabled);
    return stored === null ? false : stored === 'true';
  }

  set keyboardShortcutsEnabled(value: boolean) {
    localStorage.setItem(SettingsStore.keys.keyboardShortcutsEnabled, String(value));
  }

  /** In-app success/info toasts. Error toasts always show regardless. */
  get toastsEnabled(): boolean {
    const stored = localStorage.getItem(SettingsStore.keys.toastsEnabled);
    return stored === null ? true : stored === 'true';
  }

  set toastsEnabled(value: boolean) {
    localStorage.setItem(SettingsStore.keys.toastsEnabled, String(value));
  }

  /** OS notifications for terminal order statuses and chart-order fires,
   *  shown only while the window is unfocused. */
  get systemNotificationsEnabled(): boolean {
    const stored = localStorage.getItem(SettingsStore.keys.systemNotificationsEnabled);
    return stored === null ? true : stored === 'true';
  }

  set systemNotificationsEnabled(value: boolean) {
    localStorage.setItem(SettingsStore.keys.systemNotificationsEnabled, String(value));
  }

  get lastSymbol(): string | null {
    return localStorage.getItem(SettingsStore.keys.lastSymbol);
  }

  set lastSymbol(value: string | null) {
    if (value === null) {
      localStorage.removeItem(SettingsStore.keys.lastSymbol);
    } else {
      localStorage.setItem(SettingsStore.keys.lastSymbol, value);
    }
  }
}

File: apps/ios/0dteTrader/Core/Storage/SettingsStore.swift
import Foundation

/// Trade screen layout (PRD FR-10/FR-11).
enum TradeLayout: String, Codable, CaseIterable, Sendable {
    /// Layout A — chart fills the screen, floating Buy/Sell buttons.
    case fullscreen
    /// Layout B — chart on top, trade panel in the bottom portion.
    case split
}

/// UserDefaults-backed app settings: layout choice, split fraction, indicator
/// presets, disclaimer acceptance, last symbol, FaceID lock toggle.
final class SettingsStore: @unchecked Sendable {
    private let defaults: UserDefaults
    private let encoder = JSONEncoder()
    private let decoder = JSONDecoder()

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
    }

    private enum Keys {
        static let layoutMode = "settings.layoutMode"
        static let splitFraction = "settings.splitFraction"
        static let indicatorSettings = "settings.indicatorSettings"
        static let indicatorSettingsV1 = "settings.indicatorSettings.v1"
        static let chartDisplayV1 = "settings.chartDisplay.v1"
        static let twcSettings = "settings.twcSettings"
        static let optionsAnalyticsSettings = "settings.optionsAnalytics.v1"
        static let chartTradingSettings = "settings.chartTrading.v1"
        static let riskDisclaimerAccepted = "settings.riskDisclaimerAccepted"
        static let lastSymbol = "settings.lastSymbol"
        static let appLockEnabled = "settings.appLockEnabled"
        static let tradingLocked = "settings.tradingLocked"
        static let bypassOrderConfirmation = "settings.bypassOrderConfirmation"
        static let toastsEnabled = "settings.toastsEnabled"
        static let pushNotificationsEnabled = "settings.pushNotificationsEnabled"
        static let pushDeviceToken = "settings.pushDeviceToken"
    }

    private struct LegacyIndicatorSettings: Decodable {
        let smaEnabled: Bool?
        let smaPeriod: Int?
        let emaEnabled: Bool?
        let emaPeriod: Int?
        let vwapEnabled: Bool?
        let rsiEnabled: Bool?
        let rsiPeriod: Int?
        let macdEnabled: Bool?
        let macdFastPeriod: Int?
        let macdSlowPeriod: Int?
        let macdSignalPeriod: Int?
        let bollingerEnabled: Bool?
        let bollingerPeriod: Int?
        let bollingerMultiplier: Double?
        let volumeEnabled: Bool?
        let stochEnabled: Bool?
        let stochKPeriod: Int?
        let stochKSmooth: Int?
        let stochDPeriod: Int?
        let atrEnabled: Bool?
        let atrPeriod: Int?
    }

    func loadIndicatorSettings(registry: IndicatorRegistry) throws -> IndicatorSettingsState {
        if let data = defaults.data(forKey: Keys.indicatorSettingsV1) {
            let state = try decoder.decode(IndicatorSettingsState.self, from: data)
            try IndicatorSettingsValidator.validate(state, registry: registry)
            try completeInterruptedMigration(state: state, registry: registry)
            return state
        }
        guard let legacyData = defaults.data(forKey: Keys.indicatorSettings) else {
            let state = try IndicatorSettingsState.defaults(for: registry)
            try persistIndicatorSettings(state)
            if defaults.data(forKey: Keys.chartDisplayV1) == nil {
                try persistChartDisplayPreferences(.default)
            }
            return state
        }

        let legacy = try decoder.decode(LegacyIndicatorSettings.self, from: legacyData)
        var state = try IndicatorSettingsState.defaults(for: registry)
        applyLegacy(legacy, to: &state)
        try IndicatorSettingsValidator.validate(state, registry: registry)

        let display = ChartDisplayPreferences(
            volumeEnabled: legacy.volumeEnabled ?? ChartDisplayPreferences.default.volumeEnabled,
            volumeWeightedCandleWidth: ChartDisplayPreferences.default.volumeWeightedCandleWidth
        )
        let previousSettingsData = defaults.data(forKey: Keys.indicatorSettingsV1)
        let previousDisplayData = defaults.data(forKey: Keys.chartDisplayV1)
        do {
            try persistIndicatorSettings(state)
            try persistChartDisplayPreferences(display)
            _ = try readBack(state: state, display: display, registry: registry)
        } catch {
            restore(previousSettingsData, forKey: Keys.indicatorSettingsV1)
            restore(previousDisplayData, forKey: Keys.chartDisplayV1)
            throw error
        }
        defaults.removeObject(forKey: Keys.indicatorSettings)
        return state
    }

    private func applyLegacy(_ legacy: LegacyIndicatorSettings, to state: inout IndicatorSettingsState) {
        apply(enabled: legacy.smaEnabled, parameters: ["period": legacy.smaPeriod], id: "sma", to: &state)
        apply(enabled: legacy.emaEnabled, parameters: ["period": legacy.emaPeriod], id: "ema", to: &state)
        apply(
            enabled: legacy.vwapEnabled,
            parameters: [String: Int?](),
            id: "anchored_vwap",
            to: &state
        )
        apply(enabled: legacy.rsiEnabled, parameters: ["period": legacy.rsiPeriod], id: "rsi", to: &state)
        apply(
            enabled: legacy.macdEnabled,
            parameters: [
                "fastPeriod": legacy.macdFastPeriod,
                "slowPeriod": legacy.macdSlowPeriod,
                "signalPeriod": legacy.macdSignalPeriod,
            ],
            id: "macd",
            to: &state
        )
        apply(
            enabled: legacy.bollingerEnabled,
            parameters: [
                "period": legacy.bollingerPeriod.map(Double.init),
                "multiplier": legacy.bollingerMultiplier,
            ],
            id: "bollinger",
            to: &state
        )
        apply(
            enabled: legacy.stochEnabled,
            parameters: [
                "kPeriod": legacy.stochKPeriod,
                "kSmooth": legacy.stochKSmooth,
                "dPeriod": legacy.stochDPeriod,
            ],
            id: "stochastic",
            to: &state
        )
        apply(enabled: legacy.atrEnabled, parameters: ["period": legacy.atrPeriod], id: "atr", to: &state)
    }

    private func apply(
        enabled: Bool?,
        parameters: [String: Int?],
        id: String,
        to state: inout IndicatorSettingsState
    ) {
        if let enabled { state.indicators[id]?.enabled = enabled }
        for (parameterId, value) in parameters {
            if let value { state.indicators[id]?.parameters[parameterId] = Double(value) }
        }
    }

    private func apply(
        enabled: Bool?,
        parameters: [String: Double?],
        id: String,
        to state: inout IndicatorSettingsState
    ) {
        if let enabled { state.indicators[id]?.enabled = enabled }
        for (parameterId, value) in parameters {
            if let value { state.indicators[id]?.parameters[parameterId] = value }
        }
    }

    private func completeInterruptedMigration(
        state: IndicatorSettingsState,
        registry: IndicatorRegistry
    ) throws {
        guard let legacyData = defaults.data(forKey: Keys.indicatorSettings) else { return }
        let previousDisplayData = defaults.data(forKey: Keys.chartDisplayV1)
        let display: ChartDisplayPreferences
        if let previousDisplayData,
           let decoded = try? decoder.decode(ChartDisplayPreferences.self, from: previousDisplayData) {
            display = decoded
        } else {
            let legacy = try? decoder.decode(LegacyIndicatorSettings.self, from: legacyData)
            display = ChartDisplayPreferences(
                volumeEnabled: legacy?.volumeEnabled ?? ChartDisplayPreferences.default.volumeEnabled,
                volumeWeightedCandleWidth: ChartDisplayPreferences.default.volumeWeightedCandleWidth
            )
            do {
                try persistChartDisplayPreferences(display)
            } catch {
                restore(previousDisplayData, forKey: Keys.chartDisplayV1)
                throw error
            }
        }
        do {
            _ = try readBack(state: state, display: display, registry: registry)
            defaults.removeObject(forKey: Keys.indicatorSettings)
        } catch {
            restore(previousDisplayData, forKey: Keys.chartDisplayV1)
            throw error
        }
    }

    func updateIndicatorSettings(_ candidate: IndicatorSettingsState, registry: IndicatorRegistry) throws {
        try IndicatorSettingsValidator.validate(candidate, registry: registry)
        try persistIndicatorSettings(candidate)
        let persisted = try decoder.decode(
            IndicatorSettingsState.self,
            from: defaults.data(forKey: Keys.indicatorSettingsV1) ?? Data()
        )
        try IndicatorSettingsValidator.validate(persisted, registry: registry)
        guard persisted == candidate else {
            throw IndicatorSettingsValidationError.invalid("Indicator settings could not be verified.")
        }
    }

    func loadChartDisplayPreferences() throws -> ChartDisplayPreferences {
        guard let data = defaults.data(forKey: Keys.chartDisplayV1) else { return .default }
        return try decoder.decode(ChartDisplayPreferences.self, from: data)
    }

    func updateChartDisplayPreferences(_ preferences: ChartDisplayPreferences) throws {
        try persistChartDisplayPreferences(preferences)
        let persisted = try decoder.decode(
            ChartDisplayPreferences.self,
            from: defaults.data(forKey: Keys.chartDisplayV1) ?? Data()
        )
        guard persisted == preferences else {
            throw IndicatorSettingsValidationError.invalid("Chart display settings could not be verified.")
        }
    }

    private func persistIndicatorSettings(_ state: IndicatorSettingsState) throws {
        defaults.set(try encoder.encode(state), forKey: Keys.indicatorSettingsV1)
    }

    private func persistChartDisplayPreferences(_ display: ChartDisplayPreferences) throws {
        defaults.set(try encoder.encode(display), forKey: Keys.chartDisplayV1)
    }

    private func restore(_ data: Data?, forKey key: String) {
        if let data {
            defaults.set(data, forKey: key)
        } else {
            defaults.removeObject(forKey: key)
        }
    }

    private func readBack(
        state: IndicatorSettingsState,
        display: ChartDisplayPreferences,
        registry: IndicatorRegistry
    ) throws -> Bool {
        guard let settingsData = defaults.data(forKey: Keys.indicatorSettingsV1),
              let displayData = defaults.data(forKey: Keys.chartDisplayV1)
        else {
            throw IndicatorSettingsValidationError.invalid("Migrated settings could not be read back.")
        }
        let savedState = try decoder.decode(IndicatorSettingsState.self, from: settingsData)
        let savedDisplay = try decoder.decode(ChartDisplayPreferences.self, from: displayData)
        try IndicatorSettingsValidator.validate(savedState, registry: registry)
        guard savedState == state, savedDisplay == display else {
            throw IndicatorSettingsValidationError.invalid("Migrated settings could not be verified.")
        }
        return true
    }

    /// Layout choice persists across launches (FR-12). Defaults to split view.
    var layoutMode: TradeLayout {
        get {
            defaults.string(forKey: Keys.layoutMode)
                .flatMap(TradeLayout.init(rawValue:)) ?? .split
        }
        set { defaults.set(newValue.rawValue, forKey: Keys.layoutMode) }
    }

    /// Trade panel height as a fraction of screen height, clamped so the panel
    /// always fits the trade ticket (floor 0.32, PRD ceiling 1/2).
    var splitFraction: Double {
        get {
            let stored = defaults.double(forKey: Keys.splitFraction)
            guard stored > 0 else { return 0.38 }
            return min(0.5, max(0.32, stored))
        }
        set { defaults.set(newValue, forKey: Keys.splitFraction) }
    }

    var twcSettings: TwcHeatmapSettings {
        get {
            guard let data = defaults.data(forKey: Keys.twcSettings),
                  let settings = try? decoder.decode(TwcHeatmapSettings.self, from: data)
            else {
                return .default
            }
            return settings
        }
        set {
            if let data = try? encoder.encode(newValue) {
                defaults.set(data, forKey: Keys.twcSettings)
            }
        }
    }

    var optionsAnalyticsSettings: OptionsAnalyticsSettings {
        get {
            guard let data = defaults.data(forKey: Keys.optionsAnalyticsSettings),
                  let settings = try? decoder.decode(OptionsAnalyticsSettings.self, from: data)
            else {
                return .default
            }
            return settings
        }
        set {
            if let data = try? encoder.encode(newValue) {
                defaults.set(data, forKey: Keys.optionsAnalyticsSettings)
            }
        }
    }

    /// Chart trading (order lines drawn directly on the candles).
    var chartTradingSettings: ChartTradingSettings {
        get {
            guard let data = defaults.data(forKey: Keys.chartTradingSettings),
                  let settings = try? decoder.decode(ChartTradingSettings.self, from: data)
            else {
                return .default
            }
            return settings
        }
        set {
            if let data = try? encoder.encode(newValue) {
                defaults.set(data, forKey: Keys.chartTradingSettings)
            }
        }
    }

    var hasAcceptedRiskDisclaimer: Bool {
        get { defaults.bool(forKey: Keys.riskDisclaimerAccepted) }
        set { defaults.set(newValue, forKey: Keys.riskDisclaimerAccepted) }
    }

    var lastSymbol: String? {
        get { defaults.string(forKey: Keys.lastSymbol) }
        set { defaults.set(newValue, forKey: Keys.lastSymbol) }
    }

    var appLockEnabled: Bool {
        get { defaults.bool(forKey: Keys.appLockEnabled) }
        set { defaults.set(newValue, forKey: Keys.appLockEnabled) }
    }

    /// Trading lock: when true, every order-placing control is disabled. Persists
    /// across launches (the lock is remembered, like the layout choice). Distinct
    /// from `appLockEnabled`, which is the Face-ID app-open gate.
    var tradingLocked: Bool {
        get { defaults.bool(forKey: Keys.tradingLocked) }
        set { defaults.set(newValue, forKey: Keys.tradingLocked) }
    }

    /// Skip the buy/sell confirmation sheet and submit immediately. Per-device.
    var bypassOrderConfirmation: Bool {
        get { defaults.bool(forKey: Keys.bypassOrderConfirmation) }
        set { defaults.set(newValue, forKey: Keys.bypassOrderConfirmation) }
    }

    /// Success/info toast banners (default on). Gates those styles only —
    /// error toasts always show.
    var toastsEnabled: Bool {
        get {
            guard defaults.object(forKey: Keys.toastsEnabled) != nil else { return true }
            return defaults.bool(forKey: Keys.toastsEnabled)
        }
        set { defaults.set(newValue, forKey: Keys.toastsEnabled) }
    }

    /// Push notifications for order events. Off until the user opts in.
    var pushNotificationsEnabled: Bool {
        get { defaults.bool(forKey: Keys.pushNotificationsEnabled) }
        set { defaults.set(newValue, forKey: Keys.pushNotificationsEnabled) }
    }

    /// The APNs token (lowercase hex) last uploaded to a given server, kept
    /// so a later DELETE can retry even after a relaunch. Keyed PER SERVER:
    /// the device token is device-scoped and may be registered with several
    /// backends at once, and each registration needs its own retry handle —
    /// a single shared slot loses the old server's handle on a server switch.
    func pushDeviceToken(server: String) -> String? {
        defaults.string(forKey: "\(Keys.pushDeviceToken).\(server)")
    }

    func setPushDeviceToken(_ token: String?, server: String) {
        let key = "\(Keys.pushDeviceToken).\(server)"
        if let token {
            defaults.set(token, forKey: key)
        } else {
            defaults.removeObject(forKey: key)
        }
    }
}
